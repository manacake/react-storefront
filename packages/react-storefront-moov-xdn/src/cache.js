/**
 * @license
 * Copyright Â© 2017-2018 Moov Corporation.  All rights reserved.
 */

/**
 * Sets the correct response headers to configure browser and server caching
 * @param {Object} options
 * @param {Number} options.serverMaxAge The TTL for the edge caches.  Set to 0 to send a no-cache.  Omit to send no value for the client.
 * @param {Number} options.browserMaxAge The TTL for the browser's cache
 */
export function cache({ serverMaxAge, browserMaxAge }) {
  const cacheControl = []

  if (browserMaxAge === 0) {
    cacheControl.push('private, no-store, no-cache')
  } else if (browserMaxAge != null) {
    cacheControl.push(`max-age=${browserMaxAge}`)
  }

  if (serverMaxAge) {
    // remove these headers so varnish caching works correctly
    headers.removeAllHeaders('Age')
    headers.removeAllHeaders('Via')
    headers.removeAllHeaders('Expires')
    headers.header('X-Moov-Cache', 'true')
    cacheControl.push(`s-maxage=${serverMaxAge}`)
  }

  if (cacheControl.length) {
    headers.header('Cache-Control', cacheControl.join(', '))
  }
}

/**
 * Sets a cache time of one day for all image and font assets which are proxied from upstream.
 * @param {String} pathname
 */
export function cacheProxiedAssets(pathname, { serverMaxAge }) {
  if (pathname.match(/(jpeg|jpg|png|gif|svg|woff2?|ttf|otf)$/)) {
    cache({ serverMaxAge })
  }
}

/**
 * The TTL for far-future cached assets
 */
export const FAR_FUTURE = 290304000

/**
 * The number of seconds in a day
 */
export const ONE_DAY = 86400
