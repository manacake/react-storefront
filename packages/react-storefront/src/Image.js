/**
 * @license
 * Copyright Â© 2017-2018 Moov Corporation.  All rights reserved.
 */
import React, { Component, createRef } from 'react'
import withStyles from '@material-ui/core/styles/withStyles'
import PropTypes from 'prop-types'
import { inject } from 'mobx-react'
import classnames from 'classnames'
import VisibilitySensor from 'react-visibility-sensor'
import qs from 'qs'

export const styles = theme => ({
  root: {
    position: 'relative',
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    // Without a minimum height and width, the container will not fire
    // the visibility change
    minHeight: 1,
    minWidth: 1
  },
  fit: {
    position: 'absolute',
    top: 0,
    left: 0,
    right: 0,
    bottom: 0,
    display: 'block',
    width: '100%',
    height: '100%'
  },
  contain: {
    '& img': {
      objectFit: 'contain',
      maxHeight: '100%',
      maxWidth: '100%'
    }
  },
  fill: {
    '& img': {
      display: 'block',
      objectFit: 'contain',
      maxHeight: '100%',
      maxWidth: '100%',
      width: '100%',
      height: '100%'
    }
  }
})

/**
 * The old image optimizer intented for mobile phones only
 */
export const LEGACY_IMAGE_OPTIMIZER = 'https://opt.moovweb.net'

/**
 * The new image optimizer which supports all device types.
 */
export const RESPONSIVE_IMAGE_OPTIMIZER = 'https://optimize.moovweb.net/img'

/**
 * The new image optimizer running on localhost:8090
 */
export const LOCAL_IMAGE_OPTIMIZER = 'http://localhost:8090/img'

let optimizerUrlBase = LEGACY_IMAGE_OPTIMIZER

/**
 * Changes the base URL for the image optimizer service.
 * @param {String} url Use LEGACY_IMAGE_OPTIMIZER or RESPONSIVE_IMAGE_OPTIMIZER.
 */
export function setImageOptimizer(url) {
  optimizerUrlBase = url
}

/**
 * Creates an optimized image URL
 * @param {String} src
 * @param {Object} options
 * @param {Number} options.height The max height of the image when served to phones
 * @param {Number} options.width The max width of the image when served to phones
 * @param {Number} options.quality A number from 1-100 representing the amount to downscale the source image when served to phones
 * @param {Object} options.format "jpeg" or "webp" If webp is specified, webp will only be served to browsers that support it.
 * @return {String}
 */
export function createOptimizedSrc(src, options = {}) {
  // The legacy service used fmt, so we kept it for the new responsive service,
  // but we use "format" here because non of the other options are abbreviated
  if (options.format) options.fmt = options.format

  if (Object.keys(options).length > 0) {
    const options = { ...options, img: src }
    return `${optimizerUrlBase}/?${qs.stringify(options)}`
  } else {
    return src
  }
}

/**
 * Provide amp-compatible mobile-optimized images that can be made to auto-scale to fit the parent element
 * by setting the `fill` prop, or grow/shrink while maintaining a given aspect ratio
 * by setting the `aspectRatio` prop.
 */
@withStyles(styles, { name: 'RSFImage' })
@inject(({ app }) => ({ amp: app.amp }))
export default class Image extends Component {
  static propTypes = {
    /**
     * The URL for the image
     */
    src: PropTypes.string,

    /**
     * The URL of the image to use in case the primary image fails to load
     */
    notFoundSrc: PropTypes.string,

    /**
     * The ratio of height/width as a float.  For example: 1 when the height and width match,
     * 0.5 when height is half of the width.
     */
    aspectRatio: PropTypes.number,

    /**
     * The quality of image to retreive from 0 to 100
     */
    quality: PropTypes.number,

    /**
     * Set to true to apply object-fit:contain to the image so that it automatically
     * fits within the element's height and width.
     */
    contain: PropTypes.bool,

    /**
     * The same as contain, except images are stretched to fill the element's height and width.
     */
    fill: PropTypes.bool,

    /**
     * Set to true to wait until the image enters the viewport before loading it.
     */
    lazy: PropTypes.bool,

    /**
     * Sets the minimum amount of pixels the image can be scrolled out of view before it
     * is lazy loaded.  Defaults to 100.  You must set `lazy` in order for this setting to take effect.
     */
    lazyOffset: PropTypes.number,

    /**
     * When specified, the image will be optimized for mobile devices by the Moovweb CDN.  Accepts the following keys:
     *
     * - quality  - A number or string containing the number for the desired quality, on a scale from 1 (worst) to 100 (best).
     * - width - A number or string containing the number for the desired pixel width.
     * - height - A number or string containing the number for the desired pixel height.
     * - format - A string containing the desired file format. Accepts "webp" or "jpeg".
     */
    optimize: PropTypes.shape({
      quality: PropTypes.number,
      width: PropTypes.number,
      height: PropTypes.number,
      format: PropTypes.oneOf(['webp', 'jpeg'])
    })
  }

  static defaultProps = {
    quality: null,
    contain: false,
    fill: false,
    lazy: false,
    lazyOffset: 100,
    optimize: {}
  }

  constructor({ lazy, amp }) {
    super()

    this.state = {
      loaded: !lazy || amp,
      primaryNotFound: false
    }

    this.ref = createRef()
  }

  componentDidMount() {
    const img = this.ref.current

    if (img && img.complete && img.naturalWidth === 0) {
      this.handleNotFound()
    }
  }

  render() {
    let {
      lazy,
      lazyOffset,
      notFoundSrc,
      height,
      width,
      quality,
      amp,
      fill,
      contain,
      classes,
      className,
      aspectRatio,
      alt,
      src,
      optimize,
      ...imgAttributes
    } = this.props

    if (src == null) return null

    const { loaded, primaryNotFound } = this.state

    contain = contain || aspectRatio

    // Overiding `src` prop if `quality` was set
    src = this.getOptimizedSrc()

    if (primaryNotFound && notFoundSrc) {
      src = notFoundSrc
    }

    const assignedAttributes = {
      src,
      key: src,
      [amp ? 'class' : 'className']: classnames({
        [classes.fit]: aspectRatio != null
      }),
      layout: amp ? this.ampLayout() : null,
      height,
      width,
      alt,
      'amp-bind': imgAttributes['amp-bind']
    }

    let result = (
      <div
        className={classnames(className, {
          [classes.root]: true,
          [classes.contain]: contain,
          [classes.fill]: fill
        })}
      >
        {aspectRatio && <div style={{ paddingTop: `${aspectRatio}%` }} />}
        {amp ? (
          <amp-img {...assignedAttributes} />
        ) : (
          loaded && (
            <img
              ref={this.ref}
              {...assignedAttributes}
              {...imgAttributes}
              onError={this.handleNotFound}
            />
          )
        )}
      </div>
    )

    if (!amp && lazy) {
      result = (
        <VisibilitySensor
          active={!loaded}
          onChange={this.lazyLoad}
          partialVisibility
          offset={{ top: -lazyOffset, bottom: -lazyOffset }}
        >
          {result}
        </VisibilitySensor>
      )
    }

    return result
  }

  ampLayout() {
    const { fill, contain, aspectRatio } = this.props

    if (contain || fill || aspectRatio) {
      return 'fill'
    } else {
      return 'intrinsic'
    }
  }

  handleNotFound = () => {
    this.setState({ primaryNotFound: true })
  }

  lazyLoad = visible => {
    if (!this.state.loaded && visible) {
      this.setState({ loaded: true })
    }
  }

  getOptimizedSrc() {
    const { src, quality, optimize } = this.props
    return createOptimizedSrc(src, { ...optimize, quality })
  }
}
