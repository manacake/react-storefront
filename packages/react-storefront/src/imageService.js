/**
 * @license
 * Copyright Â© 2017-2019 Moov Corporation.  All rights reserved.
 */
import qs from 'qs'

/**
 * The old image optimizer intented for mobile phones only
 */
export const LEGACY_IMAGE_SERVICE = 'https://opt.moovweb.net'

/**
 * The new image optimizer which supports all device types.
 */
export const XDN_IMAGE_SERVICE = 'https://optimize.moovweb.net'

/**
 * The new image optimizer running on localhost:8090
 */
export const LOCAL_IMAGE_SERVICE = 'http://localhost:8090'

let optimizerUrlBase = LEGACY_IMAGE_SERVICE

/**
 * Changes the base URL for the Moovweb image service.
 * @param {String} url Use `LEGACY_IMAGE_OPTIMIZER` or `RESPONSIVE_IMAGE_OPTIMIZER`.
 * @param {Object} options
 * @param {Object} options.version The version of the image optimizer API to use. Defaults to '1'
 */
export function configureImageService(url, { version = '1' } = {}) {
  optimizerUrlBase = url

  if (url === XDN_IMAGE_SERVICE) {
    optimizerUrlBase += `/v${version}`
  }
}

/**
 * Resets the image service to the default implementation
 */
export function resetImageService() {
  optimizerUrlBase = LEGACY_IMAGE_SERVICE
}

/**
 * Creates an optimized image URL
 * @param {String} src
 * @param {Object} options
 * @param {Number} options.height The max height of the image when served to phones
 * @param {Number} options.width The max width of the image when served to phones
 * @param {Number} options.quality A number from 1-100 representing the amount to downscale the source image when served to phones
 * @param {Object} options.format "jpeg" or "webp" If webp is specified, webp will only be served to browsers that support it.
 * @return {String}
 */
export function createOptimizedSrc(src, options = {}) {
  // The legacy service used fmt, so we kept it for the new responsive service,
  // but we use "format" here because non of the other options are abbreviated
  if (options.format) options.fmt = options.format

  if (Object.keys(options).length > 0) {
    return `${optimizerUrlBase}/img?${qs.stringify({ ...options, img: src })}`
  } else {
    return src
  }
}
