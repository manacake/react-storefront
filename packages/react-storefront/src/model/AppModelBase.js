/**
 * @license
 * Copyright Â© 2017-2018 Moov Corporation.  All rights reserved.
 */
import { types } from 'mobx-state-tree'
import { MenuModel } from '../Menu'
import { TabsModel } from '../NavTabs'
import CategoryModelBase from './CategoryModelBase'
import SubcategoryModelBase from './SubcategoryModelBase'
import ProductModelBase from './ProductModelBase'
import UserModelBase from './UserModelBase'
import CartModelBase from './CartModelBase'
import SearchModelBase from './SearchModelBase'
import isEqual from 'lodash/isEqual'

/**
 * Represents a single breadcrumb
 * @class BreadcrumbModel
 */
export const BreadcrumbModel = types.model('BreadcrumbModel', {
  /**
   * The URL to link to
   * @type {String}
   * @memberof BreadcrumbModel
   * @instance
   */
  url: types.maybeNull(types.string),

  /**
   * The text for the link
   * @type {String}
   * @memberof BreadcrumbModel
   * @instance
   */
  text: types.string,

  /**
   * An object to be applied to the state tree when the breadcrumb is clicked.  The shape should match your AppModel class.
   * @type {Object}
   * @memberof BreadcrumbModel
   * @instance
   */
  state: types.frozen()
})

/**
 * Represents the browser's current location.
 * @class LocationModel
 */
export const LocationModel = types
  .model({
    /**
     * The protocol: "http" or "https".
     * @type {String}
     * @memberof LocationModel
     * @instance
     */
    protocol: 'https',
    /**
     * The hostname from the current URL.
     * @type {String}
     * @memberof LocationModel
     * @instance
     */
    hostname: types.maybeNull(types.string),
    /**
     * The path from the current URL.
     * @type {String}
     * @memberof LocationModel
     * @instance
     */
    pathname: types.string,
    /**
     * The search string from the current URL, including the leading "?".
     * @type {String}
     * @memberof LocationModel
     * @instance
     */
    search: types.string,
    /**
     * The port, typically '443'.
     * @type {String}
     * @memberof LocationModel
     * @instance
     */
    port: '443'
  })
  .views(self => ({
    /**
     * Returns the base URL for the site in the format {protocol}://{hostname}
     * @type {String}
     * @memberof LocationModel
     * @instance
     * @returns {String}
     */
    get urlBase() {
      return (
        // Since protocols sometimes have a colon at the end, we remove it if it exists and add it back manually
        self.protocol.replace(/:/, '') +
        '://' +
        self.hostname +
        (['443', '80'].includes(self.port) ? '' : `:${self.port}`)
      )
    },

    get uri() {
      return self.pathname + self.search
    },

    get href() {
      return self.urlBase + self.uri
    }
  }))

/**
 * A base class for you to extend when creating your `AppModel`, which will serve
 * as the root of the app state tree.
 *
 * Example:
 *
 * ```js
 *  // src/AppModel.js
 *  import { types } from 'mobx-state-tree'
 *  import AppModelBase from 'react-storefront/model/AppModelBase'
 *
 *  // Create the model for your app's state tree by extending AppModelBase.
 *  const AppModel = types.compose(
 *    AppModelBase,
 *    types
 *      .model('AppModel', {
 *        // your custom properties here
 *      })
 *  )
 *
 *  export default AppModel
 * ```
 *
 * @class AppModelBase
 */
const AppModelBase = types
  .model('AppModelBase', {
    /**
     * Will be `true` when the user has navigated to a new page but the viewport
     * has not yet been scrolled back to the top.  This is used internally by
     * React Storefront to keep lazy components from detecting whether or not
     * they are below the fold before the scroll position is reset.
     */
    scrollResetPending: false,
    /**
     * Will be `true` when rendering AMP, otherwise `false`.
     * @type {Boolean}
     * @memberof AppModelBase
     * @instance
     */
    amp: false,
    /**
     * Will be `true` when the user has lost their internet connection, otherwise `false`.
     * @type {Boolean}
     * @memberof AppModelBase
     * @instance
     */
    offline: false,
    /**
     * The default width to use for responsive layouts when rendering on the server.
     * @type {String}
     * @memberof AppModelBase
     * @instance
     */
    initialWidth: 'xs',
    /**
     * The top-level page component to display.  The value should be one of the keys in the `components`
     * prop in your `<Pages>` element in `src/App.js`
     * @type {String}
     * @memberof AppModelBase
     * @instance
     */
    page: types.maybeNull(types.string),
    /**
     * Sets the document title.
     * @type {String}
     * @memberof AppModelBase
     * @instance
     */
    title: types.maybeNull(types.string),
    /**
     * Sets the content for the `<meta name="description">` tag.
     * @type {String}
     * @memberof AppModelBase
     * @instance
     */
    description: types.maybeNull(types.string),
    /**
     * Will be `true` when the app is loading content from an AJAX request, otherwise `false`.
     * @type {Boolean}
     * @memberof AppModelBase
     * @instance
     */
    loading: false,
    /**
     * When an error occurs, this field will be populated with the error message.
     * @type {String}
     * @memberof AppModelBase
     * @instance
     */
    error: types.maybeNull(types.string),
    /**
     * When an error occurs, this field will be populated with the error stack trace as a string.
     * @type {String}
     * @memberof AppModelBase
     * @instance
     */
    stack: types.maybeNull(types.string),
    /**
     * State for the app's slide-in menu.
     * @type {MenuModel}
     * @memberof AppModelBase
     * @instance
     */
    menu: types.optional(MenuModel, {}),
    /**
     * State for the app's main navigation tabs.
     * @type {TabsModel}
     * @memberof AppModelBase
     * @instance
     */
    tabs: types.maybeNull(TabsModel),
    /**
     * State for a category page.
     * @type {CategoryModelBase}
     * @memberof AppModelBase
     * @instance
     */
    category: types.maybeNull(CategoryModelBase),
    /**
     * State for a subcategory page.
     * @type {SubcategoryModelBase}
     * @memberof AppModelBase
     * @instance
     */
    subcategory: types.maybeNull(SubcategoryModelBase),
    /**
     * State for a product page.
     * @type {ProductModelBase}
     * @memberof AppModelBase
     * @instance
     */
    product: types.maybeNull(ProductModelBase),
    /**
     * Partial state for the product being fetched from the back end.  Use this when
     * rendering the skeleton while loading a product from the server.
     * @type {ProductModelBase}
     * @memberof AppModelBase
     * @instance
     */
    loadingProduct: types.maybeNull(ProductModelBase),
    /**
     * Partial state for the subcategory being fetched from the back end.  Use this when
     * rendering the skeleton while loading a subcategory from the server.
     * @type {SubcategoryModelBase}
     * @memberof AppModelBase
     * @instance
     */
    loadingSubcategory: types.maybeNull(SubcategoryModelBase),
    /**
     * Partial state for the category being fetched from the back end.  Use this when
     * rendering the skeleton while loading a category from the server.
     * @type {CategoryModelBase}
     * @memberof AppModelBase
     * @instance
     */
    loadingCategory: types.maybeNull(CategoryModelBase),
    /**
     * State representing the logged in or anonymous user.
     * @type {UserModelBase}
     * @memberof AppModelBase
     * @instance
     */
    user: types.maybeNull(UserModelBase),
    /**
     * The browser's location
     * @type {LocationModel}
     * @memberof AppModelBase
     * @instance
     */
    location: types.maybeNull(LocationModel),
    /**
     * State for the search form and search results
     * @type {LocationModel}
     * @memberof AppModelBase
     * @instance
     */
    search: types.optional(SearchModelBase, {}),
    /**
     * URLs to display in the breadcrumbs
     * @type {BreadcrumbModel}
     * @memberof AppModelBase
     * @instance
     */
    breadcrumbs: types.optional(types.array(BreadcrumbModel), []),
    /**
     * The state of the cart
     * @type {CartModelBase}
     * @memberof AppModelBase
     * @instance
     */
    cart: types.optional(CartModelBase, {})
  })
  .views(self => ({
    /**
     * The `href` for the `<link rel="canonical">` tag.
     * @type {String}
     * @memberof AppModelBase
     * @instance
     */
    get canonicalURL() {
      const { protocol, hostname, pathname, search } = self.location
      return protocol + '//' + hostname + pathname.replace(/\.amp$/, '') + search
    },
    /**
     * The pathname and search from the current URL.
     * @type {String}
     * @memberof AppModelBase
     * @instance
     */
    get uri() {
      return self.location.pathname + self.location.search
    }
  }))
  .actions(self => ({
    /**
     * Updates scrollResetPending
     * @param {Boolean} pending The new value
     */
    setScrollResetPending(pending) {
      self.scrollResetPending = pending
    },

    /**
     * Clears the thumbnail being injected into the product skeleton
     * @memberof AppModelBase
     * @instance
     */
    clearProductThumbnail() {
      self.productThumbnail = null
    },

    /**
     * Applies a patch to the state tree.  The shape of the patch should match your `AppModel`.
     * @param {Object} patch The patch to apply
     * @param {String} action "PUSH", "POP", or "REPLACE".  Set to "POP" when navigating back, "REPLACE" when updating the state stored in `window.history` for the current URL, and "PUSH" when navigating forward.
     * @memberof AppModelBase
     * @instance
     */
    applyState(patch, action = 'PUSH') {
      if (action === 'POP') {
        // the user clicked the browser's back button
        patch = self.retainStateOnHistoryPop(patch) // ensure that state not corresponding to the URL is retained, for example the user and cart
        auditPatchOnPop(self, patch) // ensure that data for other pages is not changed.  This minimizes component reconciliation to boost performance

        if (patch.hasOwnProperty('page')) {
          self.page = patch.page // apply the page change first so the UI swaps to the previous page immediately
        }

        self.loading = false // if we're still loading the current page, we can hide the load mask immediately because we're going back to the previous page
        setImmediate(() => self.applyState(patch)) // apply the rest of the state change
      } else {
        // all other navigation
        const state = self.toJSON()

        for (let key in patch) {
          const value = patch[key]

          if (!isEqual(value, state[key])) {
            self[key] = value
          }
        }
      }
    },

    /**
     * Set connection status.
     * @param {Boolean} offline Set to `true` when the user has lost their internet connection.
     * @memberof AppModelBase
     * @instance
     */
    setOffline(offline) {
      self.offline = offline
    },

    /**
     * Returns the part of the state tree which should not be overwritten
     * when the user goes forward or back.  You can override this action
     * to retain additional branches of the tree.
     * @param {Object} patch The being applied in response to the user navigating back.
     * @returns {Object} A modified patch to apply to the state tree.
     * @memberof AppModelBase
     * @instance
     */
    retainStateOnHistoryPop(patch) {
      delete patch.cart
      delete patch.user
      delete patch.menu
      delete patch.tabs
      return patch
    },

    /**
     * Sets the user model.
     * @param {UserModelBase} user
     * @memberof AppModelBase
     * @instance
     */
    signIn(user) {
      self.user = user
    },

    /**
     * Clears the user model.
     * @memberof AppModelBase
     * @instance
     */
    signOut() {
      self.user = null
    },

    /**
     * Sets the user.
     * @param {UserModelBase} user
     * @memberof AppModelBase
     * @instance
     */
    setUser(user) {
      self.user = user
    },

    /**
     * Sets the `page` to "Error" and sets the `error` and `stack` properties based on
     * the provided `Error`.
     * @param {Error} e
     * @memberof AppModelBase
     * @instance
     */
    onError(e) {
      self.page = 'Error'
      self.error = e.message
      self.stack = e.stack
    }
  }))

/**
 * Removes values from the patch corresponding to existing values in the model that
 * implement forPage and for which forPage does not match the page being set in the patch.
 *
 * This helps minimize react component reconciliation by ensuring that values corresponding to
 * hidden pages are not changed when navigating back.
 *
 * @private
 * @param {AppModelBase} model The app model instance
 * @param {Object} patch The patch to be applied
 */
function auditPatchOnPop(model, patch) {
  if (patch.page) {
    for (let key in patch) {
      const current = model[key]

      if (current && current.shouldApplyPatchOnPop && !current.shouldApplyPatchOnPop(patch)) {
        delete patch[key]
      }
    }
  }
}

export default AppModelBase
