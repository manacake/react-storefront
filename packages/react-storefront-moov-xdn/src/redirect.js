/**
 * @license
 * Copyright Â© 2017-2018 Moov Corporation.  All rights reserved.
 */
import { cache } from './cache'

/**
 * Returns true if the request is a post from an amp page, otherwise false
 * @return {Boolean}
 */
function isAmpPost() {
  const { referer } = env

  return (
    referer && 
    referer.split('?')[0].endsWith('.amp') &&
    env.method.toLowerCase() === 'post'
  )
}

/**
 * Sends an http redirect.
 * @param {String} url The destination url.  Can be relative or absolute.
 * @param {Number} statusCode The status code to send.  Defaults to 301
 */
export function redirectTo(url, statusCode=301) {
  const protocol = (env.referer && env.referer.split(/:/)[0]) || 'https'

  if (!url.match(/https?:\/\//)) {
    url = `${protocol}://${env.host}${url}`
  }

  if (isAmpPost()) {
    headers.addHeader("amp-redirect-to", url)
    headers.addHeader("access-control-expose-headers", "AMP-Access-Control-Allow-Source-Origin,AMP-Redirect-To")
    headers.addHeader("amp-access-control-allow-source-origin", `${protocol}://${env.host}`)
  } else {
    headers.removeAllHeaders("Location")
    headers.addHeader("Location", url)
    headers.statusCode = statusCode
  }

  cache({ serverMaxAge: 0, browserMaxAge: 0 })
}

/**
 * Redirects the browser to the same URL but on https.
 */
export function redirectToHttps() {
  redirectTo(`https://${env.host}${env.path}`)
}