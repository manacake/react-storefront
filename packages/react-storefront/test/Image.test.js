/**
 * @license
 * Copyright Â© 2017-2018 Moov Corporation.  All rights reserved.
 */
import React from 'react'
import { mount } from 'enzyme'
import Image from '../src/Image'
import { Provider } from 'mobx-react'
import AppModelBase from '../src/model/AppModelBase'
import { configureImageService, XDN_IMAGE_SERVICE, LEGACY_IMAGE_SERVICE } from '../src/imageService'

describe('Image', () => {
  it('should render', () => {
    expect(
      mount(
        <Provider app={AppModelBase.create({ amp: false })}>
          <Image src="/foo.png" />
        </Provider>
      )
    ).toMatchSnapshot()
  })

  it('should use amp-img when rendering with amp: true', () => {
    expect(
      mount(
        <Provider app={AppModelBase.create({ amp: true })}>
          <Image src="/foo.png" />
        </Provider>
      )
    ).toMatchSnapshot()
  })

  it('should use height and width for amp-img when provided', () => {
    const wrapper = mount(
      <Provider app={AppModelBase.create({ amp: true })}>
        <Image src="/foo.png" height={50} width={100} />
      </Provider>
    )

    const ampImg = wrapper.find('amp-img')
    expect(ampImg.prop('height')).toBe(50)
    expect(ampImg.prop('width')).toBe(100)
    expect(ampImg.prop('layout')).toBe('intrinsic')
  })

  it('should use layout=fill for amp-img when aspectRatio is provided', () => {
    const wrapper = mount(
      <Provider app={AppModelBase.create({ amp: true })}>
        <Image src="/foo.png" aspectRatio={50} />
      </Provider>
    )

    const ampImg = wrapper.find('amp-img')
    expect(ampImg.prop('layout')).toBe('fill')
  })

  it('should lazy load the image when lazy==true', () => {
    const wrapper = mount(
      <Provider app={AppModelBase.create({ amp: false })}>
        <Image lazy src="/foo.png" aspectRatio={50} lazyOffset={-10000} />
      </Provider>
    )

    expect(wrapper).toMatchSnapshot()
  })

  it('should use the image optimizer when quality is set', () => {
    const wrapper = mount(
      <Provider app={AppModelBase.create({ amp: false })}>
        <Image quality={50} src="/foo.png" aspectRatio={50} />
      </Provider>
    )

    expect(wrapper).toMatchSnapshot()
  })

  it('should use the image optimizer when optimize is set', () => {
    const wrapper = mount(
      <Provider app={AppModelBase.create({ amp: false })}>
        <Image optimize={{ width: 300 }} src="/foo.png" />
      </Provider>
    )

    const img = wrapper.find('img')
    expect(img.prop('src')).toBe('https://opt.moovweb.net/img?width=300&img=%2Ffoo.png')
  })

  it('should use the not found image when the primary src fails', done => {
    const wrapper = mount(
      <Provider app={AppModelBase.create({ amp: false })}>
        <Image src="/foo.png" aspectRatio={50} notFoundSrc="/bar.png" />
      </Provider>
    )

    const image = wrapper.find('Image')

    image.setState({ primaryNotFound: true }, () => {
      const img = wrapper.find('img')
      expect(img.prop('src')).toBe('/bar.png')
      done()
    })
  })

  it('should leave the original src in place when notFoundSrc is not defined', done => {
    const wrapper = mount(
      <Provider app={AppModelBase.create({ amp: false })}>
        <Image src="/foo.png" aspectRatio={50} />
      </Provider>
    )
    const image = wrapper.find('Image')

    image.setState({ primaryNotFound: true }, () => {
      const img = wrapper.find('img')
      expect(img.prop('src')).toBe('/foo.png')
      done()
    })
  })

  describe('imageService/configureImageService', () => {
    beforeEach(() => {
      configureImageService(XDN_IMAGE_SERVICE, { version: '2' })
    })

    afterEach(() => {
      configureImageService(LEGACY_IMAGE_SERVICE)
    })

    it('should use the configured image service', () => {
      const wrapper = mount(
        <Provider app={AppModelBase.create({ amp: false })}>
          <Image src="https://images.com/foo.png" aspectRatio={50} optimize={{ width: 100 }} />
        </Provider>
      )
      expect(wrapper.find('img').prop('src')).toBe(
        'https://optimize.moovweb.net/v2/img?width=100&img=https%3A%2F%2Fimages.com%2Ffoo.png'
      )
    })
  })
})
