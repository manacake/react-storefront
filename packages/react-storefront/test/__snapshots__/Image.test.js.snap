// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image should lazy load the image when lazy==true 1`] = `
<Provider app={{...}}>
  <WithStyles(inject-Image) lazy={true} src="/foo.png" aspectRatio={50} lazyOffset={-10000}>
    <inject-Image lazy={true} src="/foo.png" aspectRatio={50} lazyOffset={-10000} classes={{...}}>
      <Image lazy={true} src="/foo.png" aspectRatio={50} lazyOffset={-10000} classes={{...}} amp={false} quality={{...}} contain={false} fill={false} optimize={{...}}>
        <VisibilitySensor active={true} onChange={[Function]} partialVisibility={true} offset={{...}} minTopValue={0} scrollCheck={false} scrollDelay={250} scrollThrottle={-1} resizeCheck={false} resizeDelay={250} resizeThrottle={-1} intervalCheck={true} intervalDelay={100} delayedCall={false} containment={{...}}>
          <div className="RSFImage-root-1 RSFImage-contain-3">
            <div style={{...}} />
          </div>
        </VisibilitySensor>
      </Image>
    </inject-Image>
  </WithStyles(inject-Image)>
</Provider>
`;

exports[`Image should render 1`] = `
<Provider app={{...}}>
  <WithStyles(inject-Image) src="/foo.png">
    <inject-Image src="/foo.png" classes={{...}}>
      <Image src="/foo.png" classes={{...}} amp={false} quality={{...}} contain={false} fill={false} lazy={false} lazyOffset={100} optimize={{...}}>
        <div className="RSFImage-root-1">
          <img src="/foo.png" className="" layout={{...}} height={[undefined]} width={[undefined]} alt={[undefined]} amp-bind={[undefined]} onError={[Function]} />
        </div>
      </Image>
    </inject-Image>
  </WithStyles(inject-Image)>
</Provider>
`;

exports[`Image should use amp-img when rendering with amp: true 1`] = `
<Provider app={{...}}>
  <WithStyles(inject-Image) src="/foo.png">
    <inject-Image src="/foo.png" classes={{...}}>
      <Image src="/foo.png" classes={{...}} amp={true} quality={{...}} contain={false} fill={false} lazy={false} lazyOffset={100} optimize={{...}}>
        <div className="RSFImage-root-1">
          <amp-img src="/foo.png" class="" layout="intrinsic" height={[undefined]} width={[undefined]} alt={[undefined]} amp-bind={[undefined]} />
        </div>
      </Image>
    </inject-Image>
  </WithStyles(inject-Image)>
</Provider>
`;

exports[`Image should use the image optimizer when quality is set 1`] = `
<Provider app={{...}}>
  <WithStyles(inject-Image) quality={50} src="/foo.png" aspectRatio={50}>
    <inject-Image quality={50} src="/foo.png" aspectRatio={50} classes={{...}}>
      <Image quality={50} src="/foo.png" aspectRatio={50} classes={{...}} amp={false} contain={false} fill={false} lazy={false} lazyOffset={100} optimize={{...}}>
        <div className="RSFImage-root-1 RSFImage-contain-3">
          <div style={{...}} />
          <img src="https://opt.moovweb.net/img?quality=50&img=%2Ffoo.png" className="RSFImage-fit-2" layout={{...}} height={[undefined]} width={[undefined]} alt={[undefined]} amp-bind={[undefined]} onError={[Function]} />
        </div>
      </Image>
    </inject-Image>
  </WithStyles(inject-Image)>
</Provider>
`;
